widgets/indicator/indicator.go

```diff
@@ -0,0 +1,244 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package Indicator is a widget that displays the status of a background operation
+// or shows whether a value is (on/off)
+package indicator
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	"math"
+	"sync"
+
+	"github.com/mum4k/termdash/align"
+	"github.com/mum4k/termdash/internal/alignfor"
+	"github.com/mum4k/termdash/internal/area"
+	"github.com/mum4k/termdash/internal/canvas"
+	"github.com/mum4k/termdash/internal/canvas/braille"
+	"github.com/mum4k/termdash/internal/draw"
+	"github.com/mum4k/termdash/internal/runewidth"
+	"github.com/mum4k/termdash/terminal/terminalapi"
+	"github.com/mum4k/termdash/widgetapi"
+)
+
+// Indicator displays the progress of an operation by filling a partial circle and
+// The circle can has a "hole" in the middle, which is where the name comes from.
+// Implements widgetapi.Widget. This object is thread-safe.
+type Indicator struct {
+	// status is the current indication that will be drawn.
+	status bool
+	// mu protects the Indicator.
+	mu sync.Mutex
+
+	// opts are the provided options.
+	opts *options
+}
+
+// New returns a new Indicator.
+func New(opts ...Option) (*Indicator, error) {
+	opt := newOptions()
+	for _, o := range opts {
+		o.set(opt)
+	}
+	if err := opt.validate(); err != nil {
+		return nil, err
+	}
+	return &Indicator{
+		opts: opt,
+	}, nil
+}
+
+func (i *Indicator) On() error {
+	i.status = true
+	return nil
+}
+
+func (i *Indicator) Off() error {
+	i.status = false
+	return nil
+}
+
+func (i *Indicator) Toggle() error {
+	if i.status == true {
+		i.status = false
+	} else {
+		i.status = true
+	}
+	return nil
+}
+
+func (i *Indicator) drawLabel(cvs *canvas.Canvas, labelAr image.Rectangle) error {
+	start, err := alignfor.Text(labelAr, i.opts.label, i.opts.labelAlign, align.VerticalMiddle)
+	if err != nil {
+		return err
+	}
+	if err := draw.Text(
+		cvs, i.opts.label, start,
+		draw.TextOverrunMode(draw.OverrunModeThreeDot),
+		draw.TextMaxX(labelAr.Max.X),
+		draw.TextCellOpts(i.opts.labelCellOpts...),
+	); err != nil {
+		return err
+	}
+	return nil
+}
+
+// innerholeRadius calculates the radius of the "inner hole" in the indicator.
+// Returns zero if no hole should be drawn.
+func (i *Indicator) innerHoleRadius(indicatorRadius int) int {
+	r := int(math.Round(float64(indicatorRadius) / 100 * float64(90)))
+	if r < 2 { // Smallest possible circle radius.
+		return 0
+	}
+	return r
+}
+
+// Draw draws the Indicator widget onto the canvas.
+// Implements widgetapi.Widget.Draw.
+func (i *Indicator) Draw(cvs *canvas.Canvas, meta *widgetapi.Meta) error {
```

Comment:
One more thing to consider - the text label is fairly close to the indicator. We might want to consider leaving at least one empty line between them, similarly to what the Donut widget does. This was changed in the following commit for reference.

https://github.com/mum4k/termdash/commit/e9b4ba36135dfb17dc27d61ab6ee21b8e3bd72b6

Let me know what you think.

Reply:
Thanks, I've included the changes you've suggested in the latest commit.